rules/aip0121/no_cycles.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0121
+
+import (
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var noCycles = &lint.MessageRule{
+	Name:   lint.NewRuleName(121, "no-cycles"),
+	OnlyIf: utils.IsResource,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		var problems []lint.Problem
+		references := make(map[*desc.FieldDescriptor]*desc.MessageDescriptor)
+
+		for _, field := range m.GetFields() {
+			if !isMutableReference(field) {
+				continue
+			}
+
+			ref := utils.GetResourceReference(field)
+			// Ignore child_type references for right now, only focus on direct
+			// references.
+			if ref.GetChildType() != "" {
+				continue
+			}
+			resMsg := utils.FindResourceMessage(ref.GetType(), m.GetFile())
+			// Skip unresolvable resource references, or those that don't resolve
+			// to a message e.g. file-level google.api.resource_defintion.
+			if resMsg == nil {
+				continue
+			}
+			references[field] = resMsg
+		}
+
+		origRes := utils.GetResource(m)
+		for origRef, refMsg := range references {
```

Comment:
this logic will detect single-deep references, but it won't detect any cycles deeper than that (e.g. a cycle of A -> B -> C -> A). I don't think there are practically many of these examples, but still something worth catching.

Could you rewrite the algorithm in a recursive fashion? you can keep a set of seen resources and pass that along to each call, then raise a problem when you find a resource that you've encountered before ( starting wit the original resource).

Reply:
Yep, I sure can.

Reply:
Done! PTAL